### PR TITLE
interfaces: validate plug/slot uniqueness

### PR DESCRIPTION
**Pull request by zyga**
This branch improves validation of changes made to the interfaces repository. In the past this was done
at `snap.Info` level, specifically via `snap.Validate` but this was ignoring additional changes caused by
adding implicit plug or slot names. This has caused a bug where the `core` snap would define some plugs
directly in the YAML file (specifically `network-bind` plug) and and than would gain implicit `network-bind` slot through the implicit mechanism. Now direct additions of plugs and slots are also enforcing the uniqueness constraint.

NOTE: This branch might only be clean to land after the new core snap is released.

### CODE REVIEW

**Comment by zyga** [comment]
This can only land after we land https://github.com/snapcore/snapd/pull/3154

File: interfaces/repo.go [hunk 1]
```diff
@@ -155,13 +157,16 @@ func (r *Repository) AddPlug(plug *Plug) error {
 	if err := i.SanitizePlug(plug); err != nil {
 		return fmt.Errorf("cannot add plug: %v", err)
 	}
-	if _, ok := r.plugs[plug.Snap.Name()][plug.Name]; ok {
-		return fmt.Errorf("cannot add plug, snap %q already has plug %q", plug.Snap.Name(), plug.Name)
+	if _, ok := r.plugs[snapName][plug.Name]; ok {
+		return fmt.Errorf("cannot add plug %q, snap %q already has a plug with that name", plug.Name, snapName)
```
Comment (niemeyer) [inline interfaces/repo.go]: From the user perspective, there's no plug or slot being _added._ I suggest reporting this as:
```
"snap %q has plugs conflicting on name %q"
"snap %q has plug and slot conflicting on name "%q"
```
and so on.

Comment (zyga) [inline interfaces/repo.go]: Aha, good point, this is just internal thing. I'll update this.

Comment (zyga) [inline interfaces/repo.go]: Done

File: interfaces/repo.go [hunk 2]
```diff
@@ -155,13 +157,16 @@ func (r *Repository) AddPlug(plug *Plug) error {
 	if err := i.SanitizePlug(plug); err != nil {
 		return fmt.Errorf("cannot add plug: %v", err)
 	}
-	if _, ok := r.plugs[plug.Snap.Name()][plug.Name]; ok {
-		return fmt.Errorf("cannot add plug, snap %q already has plug %q", plug.Snap.Name(), plug.Name)
+	if _, ok := r.plugs[snapName][plug.Name]; ok {
+		return fmt.Errorf("snap %q has plug and slot conflicting on name %q", snapName, plug.Name)
```
Comment (niemeyer) [inline interfaces/repo.go]: In this case it's plug and plug. Per the original comment here:
```
"snap %q has plugs conflicting on name %q"

File: interfaces/repo.go [hunk 3]
```diff
@@ -249,13 +256,16 @@ func (r *Repository) AddSlot(slot *Slot) error {
 	if err := i.SanitizeSlot(slot); err != nil {
 		return fmt.Errorf("cannot add slot: %v", err)
 	}
-	if _, ok := r.slots[slot.Snap.Name()][slot.Name]; ok {
-		return fmt.Errorf("cannot add slot, snap %q already has slot %q", slot.Snap.Name(), slot.Name)
+	if _, ok := r.slots[snapName][slot.Name]; ok {
+		return fmt.Errorf("snap %q has plug and slot conflicting on name %q", snapName, slot.Name)
```
Comment (niemeyer) [inline interfaces/repo.go]: Similarly:
```
"snap %q has slots conflicting on name %q"

File: interfaces/repo.go (hunk 2 above)
Comment (zyga) [inline interfaces/repo.go]: Ah, sorry, missed this detail. Same below. Correcting now.

File: interfaces/repo.go (hunk 3 above)
Comment (zyga) [inline interfaces/repo.go]: Corrected as above.

**Pull request merged**